cli/printlog/compact/compact.go

```diff
@@ -133,6 +301,9 @@ func (slr *SpawnLogReconstructor) reconstructSpawn(s *spb.ExecLogEntry_Spawn) (*
 		}
 		spawnInputs = append(spawnInputs, file)
 	}
+	if sort {
+		slices.SortFunc(spawnInputs, sortFile)
```

Comment:
I'm not entirely sure whether we want this sorting: Different orders typically result in different command lines, so I'm not sure whether this leads to cleaner diffs overall. Have you tried it?

Reply:
I sent a patch with an integration test included.

The result is something like this

```
                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -15,9 +15,2 @@
                                     {
                                -      "path": "external/bazel_tools/tools/genrule/genrule-setup.sh",
                                -      "digest": {
                                -        "hash": "ecbd7e347af9b7a4ddc4ea397b195bd0c794e1dbfd666ee29ef98dd779cf8439",
                                -        "sizeBytes": "858"
                                -      }
                                -    },
                                -    {
                                       "path": "bazel-out/stable-status.txt",
                                @@ -33,2 +26,9 @@
                                         "sizeBytes": "226"
                                +      }
                                +    },
                                +    {
                                +      "path": "external/bazel_tools/tools/genrule/genrule-setup.sh",
                                +      "digest": {
                                +        "hash": "ecbd7e347af9b7a4ddc4ea397b195bd0c794e1dbfd666ee29ef98dd779cf8439",
                                +        "sizeBytes": "858"
                                       }
                Test:           TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_false/run_0
=== RUN   TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_false/run_1
=== RUN   TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_false/run_2
--- FAIL: TestPrintCompactExec (0.48s)
    --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_true (0.05s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_true/run_0 (0.02s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_true/run_1 (0.02s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_true/run_2 (0.02s)
    --- FAIL: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_false (0.05s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_false/run_0 (0.02s)
        --- FAIL: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_false/run_1 (0.02s)
        --- FAIL: TestPrintCompactExec/testdata/bazel_compact_exec_log.binpb.zstd/isSorted_false/run_2 (0.02s)
    --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_true (0.05s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_true/run_0 (0.02s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_true/run_1 (0.02s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_true/run_2 (0.02s)
    --- FAIL: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_false (0.05s)
        --- FAIL: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_false/run_0 (0.02s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_false/run_1 (0.02s)
        --- PASS: TestPrintCompactExec/testdata/bazel_compact_exec_log_2.binpb.zstd/isSorted_false/run_2 (0.02s)
FAIL
```

Which demonstrates the flaky, nondeterministic result caused by unsorted inputs.

Reply:
https://buildbuddy.buildbuddy.io/invocation/d910dd96-ef12-44a1-bfd6-14f2b0ea6d05?target=%2F%2Fcli%2Ftest%2Fintegration%2Fcompact%3Acompact_test&targetStatus=6 here is a better view

Reply:
Realistically, we(or our users) are not consuming this output to recreate the spawn execution, so I think the input positions, relative to command args, do not matter as much.

The usage pattern here is to print the json in a deterministic way that is (a) filterable with tools like `jq` and (b) diff-friendly, which we do achieve by sorting the inputs.


Im open to other ideas though 😛

Reply:
However, I do think that sorting the ListedOutputs could be an anti-pattern. In the current sorting algorithm, the first ListedOutputs is used to compare 2 SpawnExec.

I will remove that.

Reply:
Does this mean that the order of inputs of a given spawn can change non-deterministically in the reconstructed exec log (without sorting), even if nothing about the build is changed? That sounds like a bug in the spawn log generator to me, the order should always be based on nested set order. 

How did you generate the spawn logs here?

Reply:
Ah, I found 2 bugs in this:

- In `reconstructInputs`, we were tracking the InputSet in a map which in Go, the order is unstable. Fixed this by providing a slice that tracks ordering side-by-side with the map.

- The json output of protojson is deemed to be unstable. Random spaces are added based on the hash of the current binary. So we need to pipe the json output to another json formatter to get a stable output.

Both are now fixed. PTAL.